event-web-framework/draft.markdown

```diff
@@ -0,0 +1,872 @@
+
+```lisp
+	(defmethod start ((port integer))
+	  (let ((server (socket-listen usocket:*wildcard-host* port :reuse-address t :element-type 'octet))
+		    (conns (make-hash-table)))
+	    (unwind-protect
+		    (loop (loop for ready in (wait-for-input (cons server (alexandria:hash-table-keys conns)) :ready-only t)
+			            do (process-ready ready conns)))
+	     (loop for c being the hash-keys of conns
+		     do (loop while (socket-close c)))
+	     (loop while (socket-close server)))))
+```
+
+In this loop, we have:
+
+- a server socket that listens for incoming connections;
+- a structure to store connections/buffers;
+- an infinite loop waiting for new handshakes or incoming data on an existing connection;
+- cleanup clauses to prevent dangling sockets that are unexpectedly killed (e.g. by an interrupt)
+
+If you haven't written a Common Lisp program before, this code block requires some explanation. What we have written here is a _method definition_. While Lisp is popularly known as a functional language, it also has its own system for object-oriented programming called "The Common Lisp Object System", which is usually abbreviated as "CLOS"[^CLOSpronounce]. 
+
+[^CLOSpronounce]: Pronounced "kloss", "see-loss" or "see-lows", depending on who you talk to.
+
+### CLOS and Generic Functions
+
+In CLOS, instead of focusing on classes and methods, we instead write _generic functions_[^juliachap] that are implemented as collections of _methods_. In this model, methods don't _belong_ to classes, they _specialize on_ types. The `start` method we just wrote is a unary method where the argument `port` is _specialized on_ the type `integer`. This means that we could have several implementations of `start` where `port` varies in type, and the runtime will select which implementation to use depending on the type of `port` when `start` is called.
+
+[^juliachap]: The Julia programming language takes a similar approach to object-oriented programming; you can learn more about it in FIXME.
+
+More generally, methods can specialize on more than one argument. When a `method` is called, the runtime:
+
+- dispatches on the type of its arguments to figure out which method body should be run, and
+- runs the appropriate function.
+
+You can think about it as a giant table with "Type Name" down the first column and "Operation" across the first row, as in Table FIXME.
+
+```
+            | Addition | Subtraction | Multiplication |
+    ----------------------------------------------------------
+    Integer |          |             |                |
+	-----------------------------------------------------
+	Real    |          |             |                |
+	-----------------------------------------------
+	Complex |          |             |
+```
+
+NOTE: Is this a misleading analogy? If we consider the specific case of these being binary operators (let's not handle things like (+ x y z w)) isn't it more like an N-d table, or for simplicity:
```

Comment:
Tried to merge the 500 Lines changes back into my fork so I could open
another pull request, but the push is taking an agonizingly long time :/

Have some text files instead.

The markdown is an indented and modified copy of the chapter (with some
notes handled). The asd, and lisps are some minor code changes I had to
make to avoid blocking on writes (thanks for pointing that out
incidentally).

On Sun, Mar 27, 2016 at 9:20 AM, Michael DiBernardo <
notifications@github.com> wrote:

> In event-web-framework/draft.markdown
> <https://github.com/aosabook/500lines/pull/195#discussion_r57528338>:
>
> > +- dispatches on the type of its arguments to figure out which method body should be run, and
> > +- runs the appropriate function.
> > +
> > +You can think about it as a giant table with "Type Name" down the first column and "Operation" across the first row, as in Table FIXME.
> > +
> > +```
> > +            | Addition | Subtraction | Multiplication |
> > +    ----------------------------------------------------------
> > +    Integer |          |             |                |
> > +	-----------------------------------------------------
> > +	Real    |          |             |                |
> > +	-----------------------------------------------
> > +	Complex |          |             |
> > +```
> > +
> > +NOTE: Is this a misleading analogy? If we consider the specific case of these being binary operators (let's not handle things like (+ x y z w)) isn't it more like an N-d table, or for simplicity:
>
> I decided to just skip the table entirely.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/aosabook/500lines/pull/195/files/93201bdf5b809aa0a4fe05855e30ad54c13963ec#r57528338>
>